File: typescript/graph/priority_queue.ts
interface IBinaryHeap<T> {
  elements: T[];

  isEmpty: () => boolean;
  insert: (element: T) => void;
  extractMin: () => T;
}

export interface INode<T> {
  value: T;
  priority: number;
}

interface IPriorityQueue<T> {
  elements: INode<T>[];

  isEmpty: () => boolean;
  enqueue: (element: INode<T>) => void;
  dequeue: () => INode<T> | undefined;
  swap: (i: number, j: number) => void;
  bubbleUp: () => void;
  bubbleDown: () => void;
  updatePriority: (element: INode<T>) => void;
}

class PriorityQueue<T> implements IPriorityQueue<T> {
  elements: INode<T>[];

  constructor() {
    this.elements = [];
  }

  isEmpty(): boolean {
    return this.elements.length === 0;
  }

  enqueue(element: INode<T>) {
    this.elements.push(element);
    this.bubbleUp();
  }

  dequeue(): INode<T> | undefined {
    this.swap(0, this.elements.length - 1);
    const element = this.elements.pop();
    this.bubbleDown();

    return element;
  }

  swap(i: number, j: number): void {
    const temp = this.elements[i];
    this.elements[i] = this.elements[j];
    this.elements[j] = temp;
  }

  updatePriority({ value, priority }: INode<T>) {
    const index = this.elements.findIndex((e) => e.value === value);
    this.elements.splice(index, 1);
    this.enqueue({ value, priority });
  }

  bubbleUp(): void {
    let index = this.elements.length - 1;

    while (index > 0) {
      const parentIndex = Math.floor((index - 1) / 2);
      const parentPriority = this.elements[parentIndex].priority;
      const childPriority = this.elements[index].priority;

      if (parentPriority > childPriority) {
        this.swap(index, parentIndex);
        index = parentIndex;
      } else {
        break;
      }
    }
  }

  bubbleDown(): void {
    let index = 0;
    const length = this.elements.length;
    if (length === 0) return;

    const priority = this.elements[index].priority;

    while (true) {
      let indexToSwap;
      const lChildIndex = 2 * index + 1;
      const rChildIndex = 2 * index + 2;

      const lChildPriority =
        lChildIndex < length ? this.elements[lChildIndex].priority : undefined;
      const rChildPriority =
        rChildIndex < length ? this.elements[rChildIndex].priority : undefined;

      if (lChildPriority && lChildPriority < priority) {
        indexToSwap = lChildIndex;
      }

      if (rChildPriority) {
        if (
          (!indexToSwap && rChildPriority < priority) ||
          (!!indexToSwap && lChildPriority && rChildPriority < lChildPriority)
        ) {
          indexToSwap = rChildIndex;
        }
      }

      if (!indexToSwap) break;

      this.swap(index, indexToSwap);
      index = indexToSwap;
    }
  }
}

export default PriorityQueue;

File: dart/lib/collections/priority_queue.dart
class Node<T> {
  T value;
  int priority;

  Node(this.value, this.priority);

  Node<R>? cast<R>() => value is R ? Node(value as R, priority) : null;

  @override
  bool operator ==(Object other) =>
      other is Node && other.value == value && other.priority == priority;

  @override
  String toString() => 'Node<$value, $priority>';
}

class PriorityQueue<T> {
  List<Node<T>> elements = [];

  bool isEmpty() {
    return elements.isEmpty;
  }

  void enqueue(Node<T> element) {
    elements.add(element);
    bubbleUp();
  }

  Node<T>? dequeue() {
    swap(0, elements.length - 1);
    final element = elements.removeLast();
    bubbleDown();

    return element;
  }

  void updatePriority(Node<T> node) {
    final index = elements.indexWhere((e) => e.value == node.value);
    elements.removeAt(index);
    enqueue(node);
  }

  void bubbleUp() {
    var index = elements.length - 1;

    while (index > 0) {
      final parentIndex = (index - 1) ~/ 2;
      final parentPriority = elements[parentIndex].priority;
      final childPriority = elements[index].priority;

      if (parentPriority > childPriority) {
        swap(index, parentIndex);
        index = parentIndex;
      } else {
        break;
      }
    }
  }

  void bubbleDown() {
    if (elements.isEmpty) {
      return;
    }

    var index = 0;
    final length = elements.length;
    final priority = elements[index].priority;

    while (true) {
      var indexToSwap;
      final lIndex = 2 * index + 1;
      final rIndex = 2 * index + 2;
      final lPriority = lIndex < length ? elements[lIndex].priority : null;
      final rPriority = rIndex < length ? elements[rIndex].priority : null;

      if (lPriority != null && lPriority < priority) {
        indexToSwap = lIndex;
      }

      if (rPriority != null) {
        if ((indexToSwap == null && rPriority < priority) ||
            (indexToSwap != null &&
                lPriority != null &&
                rPriority < lPriority)) {
          indexToSwap = rIndex;
        }
      }

      if (indexToSwap == null) break;

      swap(index, indexToSwap);
      index = indexToSwap;
    }
  }

  void swap(int i, int j) {
    final tmp = elements[i];
    elements[i] = elements[j];
    elements[j] = tmp;
  }
}
